dlpy/model.py

```diff
@@ -679,6 +679,362 @@ def train(self, table, attributes=None, inputs=None, nominals=None, texts=None,
+            training data samples.
+            Valid Values: 'ascending', 'descending', 'hash'
+            Default : 'ascending'
+        train_from_scratch : bool, optional
+            When set to True, it ignores the existing weights and trains the model from the scracth.
+        tensorboard: :class:`TensorBoard`, optional
+            Specifies the TensorBoard instance to log model training statistics.
+            Default : None
+        Returns
+        --------
+        :class:`CASResults`
+
+        """
+        # set reference to the training and validation table
+        self.train_tbl = data
+        self.valid_tbl = valid_table
+
+        input_tbl_opts = input_table_check(data)
+        input_table = self.conn.CASTable(**input_tbl_opts)
+
+        if data_specs is None and inputs is None:
+            from dlpy.images import ImageTable
+            if isinstance(input_table, ImageTable):
+                inputs = input_table.running_image_column
+            elif '_image_' in input_table.columns.tolist():
+                print('NOTE: Inputs=_image_ is used')
+                inputs = '_image_'
+            else:
+                raise DLPyError('either dataspecs or inputs need to be non-None')
+
+        if optimizer is None:
+            optimizer = Optimizer(algorithm=VanillaSolver(learning_rate=lr),  mini_batch_size=mini_batch_size,
+                                  max_epochs=max_epochs, log_level=log_level)
+        else:
+            if not isinstance(optimizer, Optimizer):
+                raise DLPyError('optimizer should be an Optimizer object')
+
+        max_epochs = optimizer['maxepochs']
+
+        if target is None and '_label_' in input_table.columns.tolist():
+            target = '_label_'
+
+        # check whether the field is none or not
+        if self.model_weights is not None and self.model_weights.to_table_params()['name'].upper() in \
+                list(self._retrieve_('table.tableinfo').TableInfo.Name):
+            if train_from_scratch:
+                print('NOTE: Ignoring the existing weights and training from scratch.')
+                init_weights = None
+                self.n_epochs = 0
+            else:
+                print('NOTE: Training based on existing weights.')
+                init_weights = self.model_weights
+        else:
+            print('NOTE: Training from scratch.')
+            init_weights = None
+            self.n_epochs = 0
+
+        # when model_weights is none, reset it
+        if self.model_weights is None:
+            self.model_weights = self.conn.CASTable('{}_weights'.format(self.model_name))
+
+        if save_best_weights and self.best_weights is None:
+            self.best_weights = random_name('model_best_weights', 6)
+
+        if tensorboard:
+            if tensorboard.use_valid and not valid_table:
+                raise DLPyError('Cannot log validation data without a valid_table')
```

Comment:
Watch the capitalization of DLPyErrors.  Some of them you have a lower-case character at the beginning and upper-case on others.  They should be consistent.